src/nuclide.cpp

```diff
@@ -840,15 +835,14 @@ void Nuclide::calculate_urr_xs(int i_temp, Particle& p) const
   double r = future_prn(static_cast<int64_t>(index_), *p.current_seed());
   p.stream() = STREAM_TRACKING;
 
-  int i_low = 0;
-  while (urr.prob_(i_energy, URRTableParam::CUM_PROB, i_low) <= r) {
-    ++i_low;
-  };
-
-  int i_up = 0;
-  while (urr.prob_(i_energy + 1, URRTableParam::CUM_PROB, i_up) <= r) {
-    ++i_up;
-  };
+  // Warning: this assumes row-major order of cdf_values_. Also, should it
+  // really be +1?
+  int i_low = upper_bound_index(&urr.cdf_values_(i_energy, 0),
+                &urr.cdf_values_(i_energy, 0) + urr.n_cdf(), r) +
+              1;
+  int i_up = upper_bound_index(&urr.cdf_values_(i_energy + 1, 0),
+               &urr.cdf_values_(i_energy + 1, 0) + urr.n_cdf(), r) +
```

Comment:
What's the reason for upper_bound here and lower_bound above (for energies)?

Reply:
As I interpreted the previous `for` loops in the code, this should preserve the old behavior if any values are repeated in the energy or CDF arrays. In energy, you increment i_energy until the particle energy is greater than or equal to that energy. For the CDFs, you increment until `r` is greater than or equal to the CDF value.